File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        carry = 0
        head = ListNode((l1.val + l2.val + carry) % 10)
        carry = (l1.val + l2.val + carry) // 10
        l1 = l1.next
        l2 = l2.next
        curr = head
        while l1 is not None or l2 is not None:
            if (l1 is None):
                curr.next = ListNode((l2.val + carry) % 10)
                carry = (l2.val + carry) // 10
                curr = curr.next
                l2 = l2.next
            elif (l2 is None):
                curr.next = ListNode((l1.val + carry) % 10)
                carry = (l1.val + carry) // 10
                curr = curr.next
                l1 = l1.next
            else:
                curr.next = ListNode((l1.val + l2.val + carry) % 10)
                carry = (l1.val + l2.val + carry) // 10  
                curr = curr.next
                l1 = l1.next
                l2 = l2.next
        while carry > 0:
            curr.next = ListNode(carry % 10)
            carry = carry // 10  
            curr = curr.next
        return head    
        

File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbersHelper(ListNode* l1, ListNode* l2, int carry) {
        if (!l1 && !l2){
            if (carry){
                return new ListNode(carry, nullptr);
            } else {
                return nullptr;
            }
        }
        else if (!l1){
            int j = l2->val + carry;
            return new ListNode(j % 10, addTwoNumbersHelper(l1, l2->next, j / 10));
        }
        else if (!l2){
            int j = l1->val + carry;
            return new ListNode(j % 10, addTwoNumbersHelper(l1->next, l2, j / 10));
        }
        else {
            int j = (l1->val+l2->val+carry);
            return new ListNode(j % 10, addTwoNumbersHelper(l1->next, l2->next, j / 10));
        }
    }

    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2){
        return addTwoNumbersHelper(l1, l2, 0);
    }
};